### PR TITLE
added endpoints to register a new user, login with his username and p…

### PR DESCRIPTION
**Pull request by KingLouie1994**
…assword and receive a token after logging in

### CODE REVIEW

File: auth/auth-router.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  const payload = {
+    sub: user.id,
+    username: user.username
+  };
+  const options = {
+    expiresIn: "1d"
+  };
+  const token = jwt.sign(
+    payload,
+    process.env.JWT_SECRET || "thesecret",
```
Comment (evansibok) [inline auth/auth-router.js]: You can take out the "thesecret" string... Just create a .env file and add the `JWT_SECRET=thesecret` to it. The gitignore is already ignoring the dotenv file so it won't be pushed

File: auth/auth-router.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    options
+  );
+  return token;
+}
+
+router.post("/register", (req, res) => {
```
Comment (evansibok) [inline auth/auth-router.js]: You should create a middleware to validate user information when they are registering and signing up. I have one for my admin registration and login and you can use it as a template.

File: auth/auth-router.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      res.status(201).json(saved);
+    })
+    .catch(error => {
+      res.status(500).json({
+        message: "There was an error registrating this new user"
+      });
+    });
+});
+
+router.post("/login", (req, res) => {
```
Comment (evansibok) [inline auth/auth-router.js]: Validation middleware here too.

File: auth/restricted-middleware.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+const jwt = require("jsonwebtoken");
+
+module.exports = (req, res, next) => {
+  const token = req.headers.authorization;
+  if (token) {
+    jwt.verify(token, process.env.JWT_SECRET || "thesecret", (err, decoded) => {
```
Comment (evansibok) [inline auth/restricted-middleware.js]: If you have your .env file, there's no need for using the direct string "thesecret".

File: users/getAllUsers.js [hunk 5]
```diff
@@ -1,7 +1,7 @@
 const UsersDb = require('./users-model');
 
 function getAllUsers(req, res) {
-  UsersDb.findAllUsers()
+  UsersDb.find()
```
Comment (evansibok) [inline users/getAllUsers.js]: Your controller is missing a catch block in case getting all users don't work.

File: users/getUserById.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+function getUserById(req, res) {
+  const { id } = req.params;
+  Users.findById(id)
+    .then(user => {
+      if (user) {
+        res.status(200).json(user);
+      } else {
+        res.status(404).json({
+          message: "Couldn't find user with this id"
```
Comment (evansibok) [inline users/getUserById.js]: The application client doesn't need to know that you're trying to get the user using Id

File: users/users-model.js [hunk 7]
```diff
@@ -1,9 +1,31 @@
...
+  return db("users").where(filter);
+}
+
+function findById(id) {
+  return db("users")
+    .where({ id })
+    .first();
+}
+
+function add(user) {
```
Comment (evansibok) [inline users/users-model.js:17]: Perfect!

**Review by evansibok — CHANGES_REQUESTED** [review]
Look at my comments and make the changes. Great Job!

File: auth/auth-router.js (hunk 1 above)
Comment (KingLouie1994) [reply auth/auth-router.js]: Done

File: auth/auth-router.js (hunk 2 above)
Comment (KingLouie1994) [reply auth/auth-router.js]: Done

File: auth/auth-router.js (hunk 3 above)
Comment (KingLouie1994) [reply auth/auth-router.js]: Done

File: auth/restricted-middleware.js (hunk 4 above)
Comment (KingLouie1994) [reply auth/restricted-middleware.js]: Done

File: users/getUserById.js (hunk 6 above)
Comment (KingLouie1994) [reply users/getUserById.js]: Done

File: users/getAllUsers.js (hunk 5 above)
Comment (KingLouie1994) [reply users/getAllUsers.js]: Done in last commit

File: middlewares/validateUser.js [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  if (Object.keys(user).length === 0) {
+    res.status(400).json({
+      message: "Invalid user credentials!"
+    });
+  } else if (
+    user.full_name === "" ||
+    user.username === "" ||
+    user.email === "" ||
+    user.password === "" ||
+    user.password === ""
```
Comment (evansibok) [inline middlewares/validateUser.js]: You have two conditionals for user.password here.

**Review by evansibok — CHANGES_REQUESTED** [review]
Almost done! Just change the little bit I mentioned.

**Review by evansibok — APPROVED** [review]
Good job!

**Pull request merged**